### PR TITLE
Adding fig 5 scripts

### PR DESCRIPTION
**Pull request by sigallev**

### CODE REVIEW

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    if is_classifier:
+        y = y.astype(int)
+
+    # filter oligos that appear in less than 5% of the population
+    x = x.loc[:, x.gt(0).mean().gt(0.05)].copy()
+
+    # Add year of birth and gender
+    if metadata is None:
+        metadata = pd.read_csv(os.path.join(base_path, "cohort.csv"), index_col=0, low_memory=False)[
+            ['yob', 'gender']].dropna()
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:48]: 1. why not call read_metadata?
2. don;t need to dropna() as it is done in next line anyhow.

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    os.makedirs(os.path.dirname(output_path_prefix), exist_ok=True)
+    if overwrite or not os.path.exists(output_pickle_path):
+        model_class = GradientBoostingClassifier if is_classifier else GradientBoostingRegressor
+        clf = model_class(**params).fit(x, y)
+        explainer = shap.TreeExplainer(clf)
+        shap_values = explainer(x)
+        with open(output_pickle_path, 'wb') as f:
+            pickle.dump(shap_values, f)
+    with open(output_pickle_path, 'rb') as f:
+        shap_values = pickle.load(f)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:163]: why is this read not in else?
Any point to writing and then reading?

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Save absolute Shapley values to table
+    if overwrite or not os.path.exists(output_csv_path):
+        shap_abs_values = pd.Series(data=np.abs(explainer.shap_values(x)).sum(axis=0),
+                                    index=x.columns).sort_values(ascending=False)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:168]: explainer does not exist if os.path.exists(output_pickle_path) was already there and read.
this all needs to be together under the if in line 155

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        shap_abs_values = shap_abs_values.to_frame().rename(columns={0: ' Shapley absolute value'})
+        df_info = pd.read_csv(os.path.join(base_path, "library_contents.csv"), index_col=0, low_memory=False)[
+            ['nuc_seq', 'full name']]
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:171]: you read df_info 3 or 4 times in this code.
why not just pass it between functions?

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        oligos_shap_values = pd.merge(shap_abs_values, df_info, left_index=True, right_index=True, how='left')
+        oligos_shap_values.to_csv(output_csv_path)
+
+    try:
+        shap.plots.beeswarm(shap_values, show=False, sum_bottom_features=False, plot_size=None)
+    except Exception as e:
+        print("""You might not be using the right SHAP version.
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:178]: I am not....

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_gender.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    ax.text(-0.15, 1.0, string.ascii_lowercase[1], transform=ax.transAxes, size=14, weight='bold')
+
+    # Create sub figure c
+    ax = plot_shap(x, y, True, params, 'gender', specs=spec[2], fig=fig, overwrite=True)
+    ax.text(-0.15, 1.0, string.ascii_lowercase[2], transform=ax.transAxes, size=14, weight='bold')
+
+
+if __name__ == "__main__":
+    main()
+    plt.show()
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_gender.py:44]: why show and not savefig?

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    ax.text(-0.15, 1.0, string.ascii_lowercase[1], transform=ax.transAxes, size=14, weight='bold')
+
+    # Create sub figure c
+    ax = plot_shap(x, y, True, params, 'eggs', specs=spec[2], fig=fig, overwrite=True)
+    ax.text(-0.15, 1.0, string.ascii_lowercase[2], transform=ax.transAxes, size=14, weight='bold')
+
+
+if __name__ == "__main__":
+    main()
+    plt.show()
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py:47]: why show and not savefig?

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+from PhIPSeq_external.Analysis_allergens_and_IEDB.prediction_helpers import get_oligos, read_metadata, configure_x_y, \
+    tune_hyper_parameters, plot_correlations, draw_auc, plot_shap
+
+
+def main(tune_params=False):
+    x = get_oligos()
+    meta = read_metadata()
+    meta = meta[['yob', 'gender', 'Eats Egg Recipes']]
+    y = meta['Eats Egg Recipes']
+    meta.drop(columns=['Eats Egg Recipes'], inplace=True)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py:14]: the order of these 3 lines is wierd (and why use drop).
Why not simply:
    y = meta['Eats Egg Recipes']
    meta = meta[['yob', 'gender']]

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_gender.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    tune_hyper_parameters, plot_correlations, draw_auc, plot_shap
+
+
+def main(tune_params=False):
+    x = get_oligos()
+    meta = read_metadata()
+    meta = meta[['yob', 'gender']]
+    y = meta['gender']
+    meta.drop(columns=['gender'], inplace=True)
+    x, y = configure_x_y(x, y, meta, is_classifier=True)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_gender.py:15]: the order of these 3 lines is wierd (and why use drop).
Why not simply:
    y = meta['gender']
    meta = meta[['yob']]

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    return ax
+
+
+def main(tune_params=False):
+    x = get_oligos()
+    meta = read_metadata()
+    meta = meta[['age', 'gender']]
+    y = meta['age']
+    meta.drop(columns=['age'], inplace=True)
+    x, y = configure_x_y(x, y, meta, is_classifier=False)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:95]: the order of these 3 lines is wierd (and why use drop).
Why not simply:
y = meta['age']
meta = meta[['gender']]

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    csv_out_path = os.path.join(out_path, 'age_correlations_data.csv')
+    if overwrite or not os.path.exists(csv_out_path):
+        os.makedirs(out_path, exist_ok=True)
+        bottom_quintile = y.quantile(0.2)
+        top_quintile = y.quantile(0.8)
+        bottom_label = f"Percent of youngest quintile (<{int(bottom_quintile)}\nin whom a peptide is significantly bound"
+        top_label = f"Percent of oldest quintile (>{int(top_quintile)}\nin whom a peptide is significantly bound"
+        age_group = y.apply(
+            lambda v: bottom_label if v < bottom_quintile else (top_label if v > top_quintile else np.nan))
+        oligos = x.drop(columns=['gender']).gt(1)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:26]: why not gt(0) ?
this is log fold.... (in practive we don;t have any, but we could have had)

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        summed_df = oligos.groupby('age_group').sum().T
+        summed_df.rename(columns={k: f"{k}_1" for k in summed_df.columns}, inplace=True)
+        for col in summed_outcome.index:
+            summed_df[f"{col}_0"] = summed_outcome.loc[col] - summed_df[f"{col}_1"]
+        summed_df['oddsratio'] = summed_df.apply(
+            lambda row: fisher_exact([[row[bottom_label + '_0'], row[bottom_label + '_1']],
+                                      [row[top_label + '_0'], row[top_label + '_1']]])[0], axis=1)
+        summed_df['p-value'] = summed_df.apply(
+            lambda row: fisher_exact([[row[bottom_label + '_0'], row[bottom_label + '_1']],
+                                      [row[top_label + '_0'], row[top_label + '_1']]])[1], axis=1)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:41]: again. don;t compute fisher_exact twice (see helpers comment)

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        oligos = x.drop(columns=['gender']).gt(1)
+        oligos['age_group'] = age_group
+        oligos.dropna(inplace=True)
+
+        # Perform Fisher's exact test to identify significant oligos
+        summed_outcome = age_group.value_counts()
+        summed_df = oligos.groupby('age_group').sum().T
+        summed_df.rename(columns={k: f"{k}_1" for k in summed_df.columns}, inplace=True)
+        for col in summed_outcome.index:
+            summed_df[f"{col}_0"] = summed_outcome.loc[col] - summed_df[f"{col}_1"]
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:35]: at this point its not percentage, but whatever

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        summed_df['passed_fdr'], summed_df['fdr_corrected_p_value'] = fdrcorrection(summed_df['p-value'])
+
+        oligos = oligos.groupby('age_group').mean().T * 100.0
+        oligos = oligos.merge(summed_df[['oddsratio', 'p-value', 'passed_fdr', 'fdr_corrected_p_value']],
+                              left_index=True, right_index=True, how='left')
+        df_info = pd.read_csv(os.path.join(base_path, "library_contents.csv"), index_col=0, low_memory=False)[
+            ['nuc_seq', 'full name']]
+        oligos = oligos.merge(df_info, left_index=True, right_index=True, how='left')
+        oligos.to_csv(csv_out_path)
+    return pd.read_csv(csv_out_path, index_col=0)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:51]: why not return what we have, and only read if did not compute?
I.e.:
             return oligos
    return pd.read_csv(csv_out_path, index_col=0)

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                                      [row[top_label + '_0'], row[top_label + '_1']]])[1], axis=1)
+        summed_df['passed_fdr'], summed_df['fdr_corrected_p_value'] = fdrcorrection(summed_df['p-value'])
+
+        oligos = oligos.groupby('age_group').mean().T * 100.0
+        oligos = oligos.merge(summed_df[['oddsratio', 'p-value', 'passed_fdr', 'fdr_corrected_p_value']],
+                              left_index=True, right_index=True, how='left')
+        df_info = pd.read_csv(os.path.join(base_path, "library_contents.csv"), index_col=0, low_memory=False)[
+            ['nuc_seq', 'full name']]
+        oligos = oligos.merge(df_info, left_index=True, right_index=True, how='left')
+        oligos.to_csv(csv_out_path)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:50]: you did not sort before output this time (unlike gender and eggs)

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def make_age_correlation_subplot(x, y, ax=None):
+    if ax is None:
+        ax = plt.subplot()
+    oligos = compute_age_correlations(x, y)
+    bottom_label = [col for col in oligos.columns if '(<' in col][0]
+    top_label = [col for col in oligos.columns if '(>' in col][0]
+    sns.scatterplot(data=oligos, x=bottom_label, y=top_label, hue='passed_fdr',
+                    palette=[sns.color_palette()[0], 'black'], legend=True, ax=ax)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:61]: you did not sort by p-values, so some black dots get  written upon with blue

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    legend_handles = ax.get_legend_handles_labels()
+    ax.legend(
+        [legend_handles[0][legend_handles[1].index('True')], legend_handles[0][1 - legend_handles[1].index('True')]],
+        ['Passed FDR', 'Did not pass FDR'], title="Fisher's exact text")
+    # TODO: add annotations
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:66]: really?
wither add it, or throw the comment out

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+def make_age_prediction_figure(x, y, params, ax=None):
+    if ax is None:
+        ax = plt.subplot()
+    prediction_path = os.path.join(out_path, 'age_prediction_results.csv')
+    if not os.path.exists(prediction_path):
+        predictions_results = predict_with_cross_validation(x, y, is_classifier=False, return_predictions=True,
+                                                            **params)
+        predictions_results.to_csv(prediction_path)
+    predictions_results = pd.read_csv(prediction_path, index_col=0)
```
Comment (sigallev) [inline PhIPSeq_external/Analysis_allergens_and_IEDB/predict_age.py:78]: read only in else

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+                      left_index=True, right_index=True, how='left')
+        # Rename headers
+        df.rename(columns={True: header_true, False: header_false}, inplace=True)
+        df_info = pd.read_csv(os.path.join(base_path, "library_contents.csv"), index_col=0, low_memory=False)[
+            ['nuc_seq', 'full name']]
+        df = df.merge(df_info, left_index=True, right_index=True, how='left')
+        df.sort_values(by='fdr_corrected_p_value', inplace=True)
+        df.to_csv(output_path)
+    df = pd.read_csv(output_path, index_col=0)
+    df.sort_values(by='fdr_corrected_p_value', ascending=False, inplace=True)
```
[the remark this answers is not in the document]
Comment (kalkairis) [reply PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:226]: It's the same action but the idea is not to assume data is sorted (for example after we manually changed it in the past)

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+    df = predict_with_cross_validation(x, y, is_classifier, return_predictions=True, **model_params)
+    fpr, tpr, thresholds = roc_curve(df['y'], df['y_hat'])
+    label = f"Predictor (AUC={auc(fpr, tpr):.3f}"
+    if ax is None:
+        ax = plt.subplot()
+    if num_iterations > 0:
+        auc_confidence_interval = []
+        for _ in range(num_iterations):
+            round_fprs, round_tprs, round_thresholds = roc_curve(
+                *list(map(lambda item: item[1], train_test_split(df)[0].iteritems())))
```
[the remark this answers is not in the document]
Comment (kalkairis) [reply PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:123]: Let's discuss.

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py (hunk 3 above)
Comment (kalkairis) [reply PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:168]: this is incase a single file was created and there are issues before creating the second one. They should not be merged together.

File: PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py (hunk 4 above)
Comment (kalkairis) [reply PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py:171]: I read it only in cases were I use it, meaning if the DFs have already been computed there is not need to re-read it. That's why I don't pass it - I call it at need.

File: PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py (hunk 8 above)
Comment (kalkairis) [reply PhIPSeq_external/Analysis_allergens_and_IEDB/predict_egg_consumption.py:14]: In case we change the name of something this is more robust.

**Pull request merged**